extensions/azure/data-plane-azure-storage/src/main/java/org/eclipse/dataspaceconnector/azure/dataplane/azurestorage/pipeline/AzureStorageDataSink.java

```diff
@@ -0,0 +1,108 @@
+/*
+ *  Copyright (c) 2022 Microsoft Corporation
+ *
+ *  This program and the accompanying materials are made available under the
+ *  terms of the Apache License, Version 2.0 which is available at
+ *  https://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Contributors:
+ *       Microsoft Corporation - initial API and implementation
+ *
+ */
+package org.eclipse.dataspaceconnector.azure.dataplane.azurestorage.pipeline;
+
+import org.eclipse.dataspaceconnector.azure.dataplane.azurestorage.adapter.BlobAdapterFactory;
+import org.eclipse.dataspaceconnector.dataplane.spi.pipeline.DataSource;
+import org.eclipse.dataspaceconnector.dataplane.spi.pipeline.ParallelSink;
+import org.eclipse.dataspaceconnector.dataplane.spi.result.TransferResult;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.List;
+import java.util.Objects;
+
+import static java.lang.String.format;
+import static org.eclipse.dataspaceconnector.spi.response.ResponseStatus.ERROR_RETRY;
+
+/**
+ * Writes data into an Azure storage container.
+ */
+public class AzureStorageDataSink extends ParallelSink {
+    private String accountName;
+    private String containerName;
+    private String sharedKey;
+    private BlobAdapterFactory blobAdapterFactory;
+
+    /**
+     * Writes data into an Azure storage container.
+     */
+    protected TransferResult transferParts(List<DataSource.Part> parts) {
+        for (DataSource.Part part : parts) {
+            String blobName = part.name();
+            try (var input = part.openStream()) {
+                try (var output = blobAdapterFactory.getBlobAdapter(accountName, containerName, blobName, sharedKey)
+                        .getOutputStream()) {
+                    try {
+                        input.transferTo(output);
```

Comment:
Deep nesting in this method. Consider refactoring for better readability.

Reply:
+1 a refactoring here will improve readability.